### PR TITLE
Initial trivial hello-world entitlements agent

### PR DESCRIPTION
**Pull request by prdoyle**
The agent just calls a method to announce it has booted.

### CODE REVIEW

File: distribution/tools/entitlements-agent/build.gradle
Comment (rjernst) [inline distribution/tools/entitlements-agent/build.gradle:24]: I don't think this will work. The string interpolation will happen at gradle configuration time, but it needs to be delayed to runtime. In the past I've done hacks with GString like `"${-> '-javaagent:' + jar.archiveFile.get().asFile}`, but I'm sure @mark-vieira has a better way to do that lazy evaluation now.

File: libs/entitlements-runtime/src/main/java/org/elasticsearch/entitlements/runtime/api/EntitlementChecks.java
Comment (rjernst) [inline libs/entitlements-runtime/src/main/java/org/elasticsearch/entitlements/runtime/api/EntitlementChecks.java:15]: can we make this not public? It should be an impl detail?

I also would prefer it not be "hello" (sounds like a toy project), but something that can go into production like "agent booted".

File: libs/entitlements-runtime/src/main/java/module-info.java
Comment (rjernst) [inline libs/entitlements-runtime/src/main/java/module-info.java:13]: This should be a qualified export?

**Review by rjernst — COMMENTED** [review]
Why do we need 2 jars? Isn't the agent jar effectively on the boot classpath?

**Comment by prdoyle** [comment]
@rjernst - The idea was to separate the agent code from the runtime permission-checking code. The latter is always required to be present in a running Elasticsearch process, even if we do ahead-of-time instrumentation, but the former is only required if we're using an agent.

File: libs/entitlements-runtime/src/main/java/module-info.java
Comment (prdoyle) [reply libs/entitlements-runtime/src/main/java/module-info.java:13]: Noob question - how do I avoid the Module Not Found warning when I export this to the agent?

File: libs/entitlements-runtime/src/main/java/org/elasticsearch/entitlements/runtime/api/EntitlementChecks.java
Comment (prdoyle) [reply libs/entitlements-runtime/src/main/java/org/elasticsearch/entitlements/runtime/api/EntitlementChecks.java:15]: If I selectively export it to the agent, then public is ok, right?

File: distribution/tools/entitlements-agent/build.gradle
Comment (ldematte) [reply distribution/tools/entitlements-agent/build.gradle:24]: Is this supposed to start ES with the agent? If so, how we would do that? 
I think it's a good idea to add a minimal README.md (just a few lines atm with the basic purpose of this entitlements-agent project and how to launch it).

File: libs/entitlements-runtime/src/main/java/module-info.java
Comment (prdoyle) [reply libs/entitlements-runtime/src/main/java/module-info.java:13]: Oh wait, it's another PEBCAK. Some of my modules start with `org.` and some don't.

**Review by ldematte — COMMENTED** [review]
Do we want to have a basic IT test that asserts that ES starts correctly with the agent? This will give us the IT scaffolding + a quick sanity check things are working (and keep working).
But I'm fine adding this in this PR or a separate PR.

**Comment by prdoyle** [comment]
@ldematte - no, that makes the module start its own unit tests with the agent attached. I don't have a test for ES booting with the agent.

File: distribution/tools/entitlements-agent/build.gradle
Comment (prdoyle) [reply distribution/tools/entitlements-agent/build.gradle:24]: Does it need to be delayed to runtime if it's only for the agent's own unit tests?

**Comment by ldematte** [comment]
> I don't have a test for ES booting with the agent.

I was referring more at a `gradle :run` option to start ES with the agent (e.g. an additional option to `RunTask` to do something like `gradle :run --with-entitlements-agent` or similar.
Or any other way to start ES with the new agent added to the jvm args
@mark-vieira what's the best way to do that? Can we just pass options through (like we do with `--debug-jvm`) or do we need changes to `JvmOptionsParser` and related?

**Comment by ldematte** [comment]
(I'm OK with adding that in a follow up PR too - I'd prefer slightly to do it now so this PR is nice and well-contained, but either way works for me. Same for the IT test(s), I have a slight preference to have them added now but you can raise another PR later).

File: distribution/tools/entitlements-agent/src/main/java/org/elasticsearch/entitlements/agent/Agent.java
Comment (jdconrad) [inline distribution/tools/entitlements-agent/src/main/java/org/elasticsearch/entitlements/agent/Agent.java:16]: Should we name this `EntitlementAgent` or something a bit more descriptive to avoid overloaded terms?

**Review by jdconrad — COMMENTED** [review]
@prdoyle This is looking really good, thank you! One minor comment. I will leave it to others to approve as I don't have much expertise in the build process related to gradle.

File: distribution/tools/entitlements-agent/src/main/java/org/elasticsearch/entitlements/agent/Agent.java
Comment (prdoyle) [reply distribution/tools/entitlements-agent/src/main/java/org/elasticsearch/entitlements/agent/Agent.java:16]: I don't really have an opinion here. Whatever you guys think.

Comment (ldematte) [reply distribution/tools/entitlements-agent/src/main/java/org/elasticsearch/entitlements/agent/Agent.java:16]: ++ for EntitlementAgent - if anything, it make find in Intellij easier :)

**Comment by ldematte** [comment]
Just writing this down here so we don't forget: besides an option to run it manually, it would be great if we can have a flag or something to run the whole CI tests _with_ the agent enable. Of course, optionally!
Then we can ask delivery if we can have a periodic task or something (a github tag?) that runs e.g. part1 with the agent, to know early if we are breaking things.
This is definitely a follow-up PR though! Maybe even a separate task.

**Comment by prdoyle** [comment]
It won't be long before we want our agent always running in all tests, and even in production.

File: distribution/tools/entitlement-agent/build.gradle
Comment (ldematte) [inline distribution/tools/entitlement-agent/build.gradle:24]: Have you got an answer from delivery/found out this works as intended and does not need to be dynamic?

**Review by ldematte — APPROVED** [review]
LGTM! 
As we agreed offline, the additional stuff (gradle command, IT or QA tests, etc.) we discussed here will all go in follow-up PRs.

File: distribution/tools/entitlement-agent/build.gradle
Comment (prdoyle) [reply distribution/tools/entitlement-agent/build.gradle:24]: No, I have not. Let me follow up on that before merging.

File: distribution/tools/entitlements-agent/build.gradle
Comment (prdoyle) [reply distribution/tools/entitlements-agent/build.gradle]: @rjernst - can I merge it like this? Or does it need to be changed?

File: distribution/tools/entitlement-agent/build.gradle
Comment (breskeby) [inline distribution/tools/entitlement-agent/build.gradle:27]:
```suggestion
tasks.named('jar').configure {
```

Comment (prdoyle) [reply distribution/tools/entitlement-agent/build.gradle:27]: Alright, I don't understand the nuances here... I trust you. 😂

File: distribution/tools/entitlement-agent/build.gradle
Comment (breskeby) [inline distribution/tools/entitlement-agent/build.gradle:22]: [We want to use gradle task avoidance api in our build
](https://docs.gradle.org/current/userguide/task_configuration_avoidance.html)

```suggestion
tasks.named('test').configure {
```

File: distribution/tools/entitlement-agent/build.gradle
Comment (breskeby) [inline distribution/tools/entitlement-agent/build.gradle:23]:
```suggestion
  dependsOn('jar')
```

File: distribution/tools/entitlement-agent/build.gradle
Comment (breskeby) [inline distribution/tools/entitlement-agent/build.gradle:24]:
```suggestion
  jvmArgs "-javaagent:${tasks.named('jar').flatMap{ it.archiveFile.asFile }}"
``````

**Review by breskeby — CHANGES_REQUESTED** [review]
I suggested some changes to the gradle parts of that PR

File: distribution/tools/entitlements-agent/build.gradle
Comment (prdoyle) [reply distribution/tools/entitlements-agent/build.gradle]: @breskeby - is this something you know about?

Comment (breskeby) [reply distribution/tools/entitlements-agent/build.gradle]: you applied my suggestion and I verified the fix locally seeing the agent is correctly applied to the tests and we didn't break task avoidance: https://gradle-enterprise.elastic.co/s/q7l56qcrj2cbm/performance/configuration/by-project?openProjects=WyJwcm9qZWN0LTpkaXN0cmlidXRpb246dG9vbHM6ZW50aXRsZW1lbnQtYWdlbnQiXQ#project-:distribution:tools:entitlement-agent 
👍

**Review by breskeby — APPROVED** [review]

File: distribution/tools/entitlements-agent/build.gradle
Comment (rjernst) [reply distribution/tools/entitlements-agent/build.gradle]: > can I merge it like this? Or does it need to be changed?

Two jars seem like an unnecessary distinction to me. But I'm ok with merging as-is and debating that separately.

File: distribution/tools/entitlement-agent/build.gradle
Comment (prdoyle) [reply distribution/tools/entitlement-agent/build.gradle]: Got confirmation from @breskeby that he's happy with this.

**Pull request merged**
